### PR TITLE
EndUser ongoing monitoring

### PR DESCRIPTION
**Pull request by MatanYadaev**
## **Type**
enhancement, bug_fix

___

## **Description**
- Added new schemas for active monitoring and AML hits in `end-user.schema.ts`.
- Enhanced `EndUserService` to parse and include new data fields when updating end users.
- Updated `WebhooksService` and `HookCallbackHandlerService` to handle AML hits and active monitoring updates.
- Added new fields to the database and Prisma schema to store active monitoring and AML hits data.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>end-user.schema.ts</strong><dd><code>Define Schemas for Active Monitoring and AML Hits</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/end-user/end-user.schema.ts
<li>Defined <code>ActiveMonitoringSchema</code> and <code>EndUserActiveMonitoringsSchema</code> for <br>monitoring end users.<br> <li> Defined <code>AmlHitSchema</code> and <code>EndUserAmlHitsSchema</code> for handling AML hits <br>data.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2335/files#diff-d420ec22a9fa25cd719dfecb1e764877bcb52ca5bfbe80f12e90cded1b41e678">+32/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>end-user.service.ts</strong><dd><code>Enhance EndUser Service with Active Monitoring and AML Data Parsing</code></dd></summary>
<hr>

services/workflows-service/src/end-user/end-user.service.ts
<li>Imported new schemas for use in the <code>updateById</code> method.<br> <li> Added parsing of <code>activeMonitorings</code> and <code>amlHits</code> using the new schemas <br>before updating the database.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2335/files#diff-d9518aabac6abcfb25766c40a28700d46e9a77b8f1b447f9964db91b5a2899e1">+21/-1</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>webhooks.service.ts</strong><dd><code>Update Webhooks Service to Handle AML Hits</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/webhooks/webhooks.service.ts
<li>Imported <code>EndUserService</code> for updating end user data.<br> <li> Added handling of individual AML hits by updating end user data with <br>new AML hits.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2335/files#diff-502e376b58b4b024b86f87fb27ebdde9cdae9373f7ee4cf129672611602f32db">+10/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>hook-callback-handler.service.ts</strong><dd><code>Add AML Data Handling in Hook Callback Handler Service</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/workflow/hook-callback-handler.service.ts
<li>Added new method <code>updateEndUserWithAmlData</code> to update end user data with <br>AML hits and active monitoring.<br> <li> Enhanced <code>handleHookResponse</code> to include AML data handling.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2335/files#diff-fdcf34223f85ff30a15d709b24b380b5d53c37e98cf3bcbfd255b581c919eca1">+54/-1</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>zod-schemas.ts</strong><dd><code>Extend ConfigSchema with Ongoing Monitoring Field</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/workflow/schemas/zod-schemas.ts
<li>Added optional boolean field <code>ubosOngoingMonitoring</code> to <code>ConfigSchema</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2335/files#diff-6d086b12f0d4e85ae922a595183e3edc09ed17d4affe6ea4bf99b22fe9b3e506">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>migration.sql</strong><dd><code>Database Migration to Add Monitoring and AML Columns</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/prisma/migrations/20240430113758_add_monitoring_and_aml_columns_to_end_user/migration.sql
<li>Added <code>activeMonitorings</code> and <code>amlHits</code> columns to the <code>EndUser</code> table.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2335/files#diff-3b0d0b767b1f517c1f1592ae9a6e6ccd6e1c0a758a13cbd77707b083da23bd94">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>schema.prisma</strong><dd><code>Update Prisma Schema to Include New EndUser Fields</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/prisma/schema.prisma
<li>Added <code>activeMonitorings</code> and <code>amlHits</code> JSON fields to the <code>EndUser</code> model.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2335/files#diff-79f9532cd1d8e6c67881328fa550f851c8429f183de839b32575e959cdabef26">+15/-13</a>&nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

## Summary by CodeRabbit

- **New Features**
    - Enhanced end-user profiles with monitoring and anti-money laundering (AML) tracking capabilities.
    - Introduced new validations for end-user data related to active monitorings and AML hits.
    - Updated webhook services to support AML data updates for end users.

- **Bug Fixes**
    - Adjusted default creation timestamps for UI definitions to ensure accuracy.

- **Refactor**
    - Reordered fields in the Alert Definition model for better organization and readability.

- **Documentation**
    - Updated schemas to include new fields and validations, ensuring clarity in data handling.

### CODE REVIEW

File: services/workflows-service/src/end-user/end-user.schema.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  monitoredUntil: z.string().datetime(),
+  sessionId: z.string(),
+});
+
+export const EndUserActiveMonitoringsSchema = z.array(ActiveMonitoringSchema);
+
+const SourceSchema = z.object({
+  sourceName: z.string().optional(),
+  sourceUrl: z.string().url().optional(),
+  date: z.string().nullable().optional(),
```
Comment (Omri-Levy) [inline services/workflows-service/src/end-user/end-user.schema.ts:15]: Relevant?
```suggestion
  date: z.string().datetime().nullable().optional(),
```

File: services/workflows-service/src/workflow/hook-callback-handler.service.ts [hunk 2]
```diff
@@ -39,12 +43,24 @@ export class HookCallbackHandlerService {
...
         resultDestinationPath,
         currentProjectId,
       );
+
+      const aml = data.aml as { endUserId: string; hits: unknown[] };
+
+      await this.updateEndUserWithAmlData({
+        sessionId: data.id as string,
+        amlHits: aml.hits,
+        withActiveMonitoring: workflowRuntime.config.ubosOngoingMonitoring ?? false,
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/hook-callback-handler.service.ts]: Only if you feel like it.
```suggestion
        withActiveMonitoring: !!workflowRuntime.config.ubosOngoingMonitoring,
```

File: services/workflows-service/src/workflow/schemas/zod-schemas.ts [hunk 3]
```diff
@@ -58,6 +58,7 @@ export const ConfigSchema = z
     kybOnExitAction: z.enum(['send-event', 'redirect-to-customer-portal']).optional(),
     reportConfig: z.record(z.string(), z.unknown()).optional(),
     theme: WorkflowDefinitionConfigThemeSchema.optional(),
+    ubosOngoingMonitoring: z.boolean().optional(),
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/schemas/zod-schemas.ts]: A bit more clarity that its a boolean and a hope for consistency/conventions. Again, not critical.
```suggestion
    isUbosOngoingMonitoring: z.boolean().optional(),
```

File: services/workflows-service/src/end-user/end-user.schema.ts (hunk 1 above)
Comment (MatanYadaev) [reply services/workflows-service/src/end-user/end-user.schema.ts:15]: It's a date, not a date time.

File: services/workflows-service/src/workflow/schemas/zod-schemas.ts (hunk 3 above)
Comment (MatanYadaev) [reply services/workflows-service/src/workflow/schemas/zod-schemas.ts]: Thanks

**Pull request merged**